axlearn/common/checkpointer_orbax_emergency_replicator.py

```diff
@@ -0,0 +1,548 @@
+            temp_file = epath.Path(self._local_dir) / temp_file
+            num_nodes = jax.process_count()
+            nodes_per_slice = num_nodes // num_slices
+
+            node_rank = global_state.process_id
+            my_process_index = jax.process_index()
+            proc_index_to_node_rank = ocp.multihost.runtime_to_distributed_ids()
+
+            my_in_pipeline_index = my_process_index % nodes_per_slice
+            peer_ranks = []
+            for i in range(num_slices):
+                peer_process_index = i * nodes_per_slice + my_in_pipeline_index
+                if peer_process_index != my_process_index:
+                    peer_process_rank = proc_index_to_node_rank[peer_process_index]
+                    peer_ranks.append(peer_process_rank)
+
+            logging.info("Peers for NodeRank %s: %s", node_rank, peer_ranks)
+
+            run_name = os.environ.get("HOSTNAME").split("job")[0].rstrip("-")
+
+            replicator_yaml = f"""job-name: {run_name}
+          framework: orbax
+          assume-data-parallelism: 2
+          node-rank: {node_rank}
+          nodes: {num_nodes}
+          peer-ranks: {peer_ranks}
+          backup-interval-minutes: 5"""
+
+            temp_file.write_text("\n".join([l.strip() for l in replicator_yaml.split("\n")]))
+            os.rename(temp_file, replicator_file_path)
+            if not _wait_for_file_to_disappear(replicator_file_path):
+                logging.info("The newly created replicator.yaml was not deleted in time.")
+            else:
+                logging.info("The newly created replicator.yaml was deleted, moving forward.")
+                _block_and_process_restore_dir(self._local_dir)
+
+        ckpt_cfg: Checkpointer.Config = Checkpointer.default_config()
+        # TODO(hanzhi-zhou): this `keep_last_n` may not be what users expect since non-tensor
+        # states will save when either local or persistent checkpoint will save.
+        ckpt_cfg.keep_last_n = cfg.keep_last_n
+        ckpt_cfg.keep_every_n_steps = cfg.keep_every_n_steps
+        ckpt_cfg.storage = _TFSavablesStateStorage.default_config()
+        ckpt_cfg.storage.timeout_secs = cfg.non_tensor_async_timeout_secs
+        #        ckpt_cfg.dir = os.path.join(cfg.dir, self._NON_TENSORS_PREFIX)
+        ckpt_cfg.name = "non-tensors-checkpointer"
+
+        save_policy = cfg.save_policy.instantiate()
+        local_save_policy = cfg.local_save_policy.instantiate()
+
+        # Non-tensor states must save when either local or persistent ckpt needs to be saved for
+        # restore from either to succeed.
+        def _composite_save_policy(*, step: int, evaler_summaries: dict[str, Any]):
+            return (
+                save_policy(step=step, evaler_summaries=evaler_summaries)
+                or local_save_policy(step=step, evaler_summaries=evaler_summaries)
+                or self._reached_preemption
+            )
+
+        self._composite_save_policy = _composite_save_policy
```

Comment:
not sure here, but should not our policy be non-composite and just `local_save_policy`?